networkx/generators/expanders.py

```diff
@@ -204,3 +214,196 @@ def paley_graph(p, create_using=None):
+
+    import numpy as np
+
+    if not (isinstance(n, int) and n > 0):
+        raise nx.NetworkXError("n must be a positive integer")
+
+    if not (d >= 2):
+        raise nx.NetworkXError("d must be greater than or equal to 2")
+
+    if not (isinstance(d, int) and d % 2 == 0):
+        raise nx.NetworkXError("d must be even")
+
+    if not (n - 1 >= d):
+        raise nx.NetworkXError(
+            f"There is not enough room for {d//2} independent cycles in a graph of size {n}"
+        )
+
+    G = nx.empty_graph(n, create_using)
+
+    if n < 2:
+        return G
+
+    cycles = []
+    edges = set()
+
+    # Create d / 2 cycles
+    for i in range(d // 2):
+        # Make sure the cycles are independent to have a regular graph
+        while len(edges) != (i + 1) * n:
+            # Faster than random.permutation(n) since there are only
+            # (n-1)! distinct cycles against n! permutations of size n
+            cycle = np.concatenate((np.random.permutation(n - 1), [n - 1]))
+
+            new_edges = {
+                (u, v)
+                for u, v in pairwise(cycle, cyclic=True)
+                if (u, v) not in edges and (v, u) not in edges
+            }
+            # If the new cycle has no edges in common with previous cycles
+            # then add it to the list otherwise try again
+            if len(new_edges) == n:
+                cycles.append(cycle)
+                edges.update(new_edges)
+
+    G.add_edges_from(edges)
+
+    return G
+
+
+@not_implemented_for("directed")
+@not_implemented_for("multigraph")
+def is_regular_expander(G: nx.Graph, *, epsilon=0):
+    """Determines whether the graph G is a regular expander.
+
+    More precisely, checks whether the graph is a regular $(n, d, \lambda)$-expander
+    with $\lambda$ close to the Alon-Boppana bound and given by
+    $\lambda = 2 \sqrt{d - 1} + \epsilon$. [1]_
+
+    In the case where $\epsilon = 0 $ then if the graph successfully passes the test
+    it is a Ramanujan graph. [2]_
+
+    Parameters
+    ----------
+    G : NetworkX graph
+    epsilon : int, float, default=0
+
+    Returns
+    -------
+    bool
+        Whether the given graph is a regular $(n, d, \lambda)$-expander
+        where $\lambda = 2 \sqrt{d - 1} + \epsilon$.
+
+    Examples
+    --------
+    >>> G = nx.random_regular_expander(4, 20)
+    >>> nx.is_regular_expander(G)
+    True
+
+    References
+    ----------
+    .. [1] Alon-Boppana bound, https://en.wikipedia.org/wiki/Alon%E2%80%93Boppana_bound
+    .. [2] Ramanujan graphs, https://en.wikipedia.org/wiki/Ramanujan_graph
+
+    """
+
+    import numpy as np
+    from scipy.sparse.linalg import eigsh
+
+    if not (isinstance(epsilon, (int, float)) and epsilon >= 0):
+        raise nx.NetworkXError("epsilon must be non negative")
+
+    if not nx.is_regular(G):
+        return False
+
+    d = list(G.degree)[0][1]
+
+    A = nx.adjacency_matrix(G)
+    lams = eigsh(A.asfptype(), which="LM", k=2, return_eigenvectors=False)
```

Comment:
Don't you want this to be 'LA'? to get the largest (hopefully positive) eigenvalues?  Or does it turn out you just need the largest magnitude eigenvalues?

Reply:
I need the two largest in absolute value so I guess that's 'LM' right?

Reply:
Good -- that makes me feel more comfortable with other aspects too. I've never seen a use of the largest two eigenvalues, only the largest two in magnitude. (I'm sure there are cases for the two largest, but I've never run into them.)